app/vminsert/netstorage/replication_group.go

```diff
@@ -0,0 +1,284 @@
+package netstorage
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/consts"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/fasttime"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/memory"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/storage"
+	"github.com/VictoriaMetrics/metrics"
+	xxhash "github.com/cespare/xxhash/v2"
+)
+
+type replicationGroup struct {
+	name                     string
+	storageNodes             []*storageNode
+	maxBufSizePerStorageNode int
+
+	reroutedBR         bufRows
+	reroutedBRLock     sync.Mutex
+	reroutedBRCond     *sync.Cond
+	reroutedBufMaxSize int
+
+	reroutedRowsProcessed           *metrics.Counter
+	reroutedBufWaits                *metrics.Counter
+	reroutesTotal                   *metrics.Counter
+	rerouteErrors                   *metrics.Counter
+	rowsLostTotal                   *metrics.Counter
+	rowsIncompletelyReplicatedTotal *metrics.Counter
+
+	storageNodesWG      sync.WaitGroup
+	rerouteWorkerWG     sync.WaitGroup
+	storageNodesStopCh  chan struct{}
+	rerouteWorkerStopCh chan struct{}
+}
+
+var replicationGroups map[string]*replicationGroup
+
+func initReplicationGroups() {
+	for _, rg := range replicationGroups {
+		rg.initReplicationGroup()
+	}
+}
+
+func (rg *replicationGroup) initReplicationGroup() {
+	rg.reroutedRowsProcessed = metrics.NewCounter(fmt.Sprintf(`vm_rpc_rerouted_rows_processed_total{name="vminsert",group="%s"}`, rg.name))
+	rg.reroutedBufWaits = metrics.NewCounter(fmt.Sprintf(`vm_rpc_rerouted_buf_waits_total{name="vminsert",group="%s"}`, rg.name))
+	rg.reroutesTotal = metrics.NewCounter(fmt.Sprintf(`vm_rpc_reroutes_total{name="vminsert",group="%s"}`, rg.name))
+	_ = metrics.NewGauge(fmt.Sprintf(`vm_rpc_rerouted_rows_pending{name="vminsert",group="%s"}`, rg.name), func() float64 {
+		rg.reroutedBRLock.Lock()
+		n := rg.reroutedBR.rows
+		rg.reroutedBRLock.Unlock()
+		return float64(n)
+	})
+	_ = metrics.NewGauge(fmt.Sprintf(`vm_rpc_rerouted_buf_pending_bytes{name="vminsert",group="%s"}`, rg.name), func() float64 {
+		rg.reroutedBRLock.Lock()
+		n := len(rg.reroutedBR.buf)
+		rg.reroutedBRLock.Unlock()
+		return float64(n)
+	})
+	rg.rerouteErrors = metrics.NewCounter(fmt.Sprintf(`vm_rpc_reroute_errors_total{name="vminsert",group="%s"}`, rg.name))
+	rg.rowsLostTotal = metrics.NewCounter(fmt.Sprintf(`vm_rpc_rows_lost_total{name="vminsert",group="%s"}`, rg.name))
+	rg.rowsIncompletelyReplicatedTotal = metrics.NewCounter(fmt.Sprintf(`vm_rpc_rows_incompletely_replicated_total{name="vminsert",group="%s"}`, rg.name))
+
+	rg.reroutedBRCond = sync.NewCond(&rg.reroutedBRLock)
+	rg.rerouteWorkerStopCh = make(chan struct{})
+
+	rg.maxBufSizePerStorageNode = memory.Allowed() / 8 / len(rg.storageNodes)
+	if rg.maxBufSizePerStorageNode > consts.MaxInsertPacketSize {
+		rg.maxBufSizePerStorageNode = consts.MaxInsertPacketSize
+	}
+	rg.reroutedBufMaxSize = memory.Allowed() / 16
```

Comment:
It is better from simplicity PoV calculating these limits only once and storing them in a global scope.